packages/web3-core-logger/src/index.ts

```diff
@@ -0,0 +1,110 @@
+import { CoreErrors, CoreErrorNames } from './errors';
+import { Web3PackageErrorConfig, Web3Error, Web3ErrorDetails } from './types';
+import packageVersion from './_version';
+
+export default class Web3CoreLogger {
+    private _packageErrorConfig: Web3PackageErrorConfig;
+    private _errorsCollective: { [key: string]: Web3Error };
+
+    constructor(packageErrorConfig: Web3PackageErrorConfig) {
+        this._packageErrorConfig = packageErrorConfig;
+        this._errorsCollective = CoreErrors;
+
+        // Check for name collisions
+        for (const errorName of Object.keys(CoreErrors)) {
+            if (packageErrorConfig.errors.hasOwnProperty(errorName))
+                throw this.makeError(CoreErrorNames.duplicateErrorName, {
+                    params: { duplicateError: errorName },
+                });
+        }
+
+        this._errorsCollective = {
+            ...this._errorsCollective,
+            ...packageErrorConfig.errors,
+        };
+    }
+
+    /**
+     * Creates an instace of Error using provided {web3ErrorName} and {errorDetails}
+     *
+     * @param web3ErrorName Name of error from {CoreErrors} or {Web3PackageErrorConfig}
+     * @param errorDetails Additional details to include in error message
+     * @returns Error instance
+     */
+    makeError(web3ErrorName: string, errorDetails?: Web3ErrorDetails): Error {
+        try {
+            if (!this._errorsCollective.hasOwnProperty(web3ErrorName))
+                throw this.makeError(CoreErrorNames.unsupportedError, {
+                    params: { web3ErrorName, errorDetails },
+                });
+
+            return Error(
+                this._makeErrorString({
+                    ...this._errorsCollective[web3ErrorName],
+                    ...errorDetails,
+                })
+            );
+        } catch (error) {
+            throw error;
+        }
+    }
+
+    /**
+     * Creates a string with provided error parameters
+     *
+     * @param web3Error An object containing error details
+     * @returns Error string
+     */
+    private _makeErrorString(web3Error: Web3Error): string {
```

Comment:
Currently the `error.message` (which is a `string`) looks like:

```javascript
loggerVersion: 1.0.0-alpha.0
packageName: web3-providers-http
packageVersion: 4.0.0-alpha.0
code: 1
name: invalidClientUrl
msg: Provided web3Client is an invalid HTTP(S) URL
params: {"web3Client":{}}
```

I was thinking it might be a good idea to `JSON.stringify` `web3Error` instead of piecing together an error string, so that `error.message` is parsable with `JSON.parse(error.message)` to allow the user to handle errors more effectively

Reply:
Updated